src/domain_bridge/domain_bridge.cpp

```diff
@@ -0,0 +1,218 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "domain_bridge/domain_bridge.hpp"
+
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <set>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include "domain_bridge/domain_bridge_options.hpp"
+#include "domain_bridge/topic_bridge_options.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executor.hpp"
+#include "rcutils/logging_macros.h"
+#include "rosbag2_cpp/typesupport_helpers.hpp"
+
+#include "generic_publisher.hpp"
+#include "generic_subscription.hpp"
+#include "topic_bridge.hpp"
+
+namespace domain_bridge
+{
+
+/// Implementation of \ref DomainBridge.
+class DomainBridgeImpl
+{
+public:
+  using NodeMap = std::map<std::size_t, std::shared_ptr<rclcpp::Node>>;
+
+  explicit DomainBridgeImpl(const DomainBridgeOptions & options)
+  : options_(options)
+  {}
+
+  ~DomainBridgeImpl() = default;
+
+  rclcpp::Context::SharedPtr create_context_with_domain_id(std::size_t domain_id)
+  {
+    auto context = std::make_shared<rclcpp::Context>();
+    rclcpp::InitOptions options;
+    options.auto_initialize_logging(false).set_domain_id(domain_id);
+    context->init(0, nullptr, options);
+    return context;
+  }
+
+  rclcpp::NodeOptions create_node_options(rclcpp::Context::SharedPtr context)
+  {
+    rclcpp::NodeOptions options;
+    return options.context(context)
+           .use_global_arguments(false)
+           .start_parameter_services(false)
+           .start_parameter_event_publisher(false);
+  }
+
+  rclcpp::Node::SharedPtr get_node_for_domain(std::size_t domain_id)
+  {
+    auto domain_id_node_pair = node_map_.find(domain_id);
+
+    // If we don't already have a node for the domain, create one
+    if (node_map_.end() == domain_id_node_pair) {
+      auto context = create_context_with_domain_id(domain_id);
+      auto node_options = create_node_options(context);
+      std::ostringstream oss;
+      oss << "domain_bridge_" << std::to_string(options_.id()) << "_" << std::to_string(domain_id);
+      auto node = std::make_shared<rclcpp::Node>(oss.str(), node_options);
+      node_map_[domain_id] = node;
+      return node;
+    }
+
+    return domain_id_node_pair->second;
+  }
+
+  std::shared_ptr<GenericPublisher> create_publisher(
+    rclcpp::Node::SharedPtr node,
+    const std::string & topic_name,
+    const rosidl_message_type_support_t & typesupport_handle,
+    rclcpp::CallbackGroup::SharedPtr group)
+  {
+    auto publisher = std::make_shared<GenericPublisher>(
+      node->get_node_base_interface().get(),
+      typesupport_handle,
+      topic_name,
+      rclcpp::QoS(10));
+    node->get_node_topics_interface()->add_publisher(publisher, std::move(group));
```

Comment:
this doesn't look fine IMO.
The `GenericPublisher` implementation should already be calling this internally

Reply:
That makes sense, I could move this inside the GenericPublisher. But it shouldn't break anything doing it here right? It just makes more sense to encapsulate it in the GenericPublisher.

Reply:
I guess this will be fixed when switching to the upstream GenericPublisher too.

Reply:
> But it shouldn't break anything doing it here right?

I thought it was going to break something because I imagined that `rclcpp::PublisherBase` was already doing that, but it seems that's not the case.
but `rclcpp::create_publisher` is currently doing this.

> It just makes more sense to encapsulate it in the GenericPublisher.

:+1: yes

Reply:
I just copied this implementation from rosbag2_transport, but hopefully the rclcpp version will do this encapsulation.